File: 40-combination-sum-ii/combination-sum-ii.ts
var combinationSum2 = function(candidates, target) {
    const ans = [];
    candidates.sort((a, b) => a - b); // Sort the candidates to handle duplicates
    getAllComb(candidates, 0, 0, []);

    function getAllComb(arr, i, curSum, curArr) {
        if (curSum === target) {
            ans.push([...curArr]);
            return;
        }
        if (i >= arr.length || curSum > target) {
            return;
        }

        // Include the current number
        curArr.push(arr[i]);
        getAllComb(arr, i + 1, curSum + arr[i], curArr);
        curArr.pop();

        // Skip duplicates
        while (i + 1 < arr.length && arr[i] === arr[i + 1]) {
            i++;
        }

        // Exclude the current number and move to the next unique number
        getAllComb(arr, i + 1, curSum, curArr);
    }

    return ans;
};

File: 40-combination-sum-ii/combination-sum-ii.java
class Solution { 
    public List<List<Integer>> combinationSum2(int[] candidates, int target) { 
        Arrays.sort(candidates); 
        List<List<Integer>> ans = new ArrayList<>(); 
        List<Integer> curList = new ArrayList<>(); 
        
        // Start the recursive process once from index 0
        generate(candidates, target, ans, curList, 0, 0); 
        return ans; 
    } 

    private void generate(int[] candidates, int target, List<List<Integer>> ans, List<Integer> curList, int index, int curSum) { 
        if (curSum == target) { 
            ans.add(new ArrayList<>(curList)); 
            return; 
        } 
        if (curSum > target || index >= candidates.length) {
            return; 
        }

        // Loop starts directly at 'index' to handle the current element
        for (int i = index; i < candidates.length; i++) { 
            // Correct duplicate skipping within the same decision level
            if (i > index && candidates[i] == candidates[i - 1]) {
                continue; 
            }
            
            curList.add(candidates[i]); 
            // Move to 'i + 1' to avoid reusable elements 
            generate(candidates, target, ans, curList, i + 1, curSum + candidates[i]); 
            curList.remove(curList.size() - 1); 
        } 
    } 
}